// git-watcher/GitParser.js
module.exports = {
	
	getStatus: function(modulePath, callback) {
		var me = this;
		require('child_process').exec('git status --porcelain -z', {cwd: modulePath}, function(err, stdout, stderr) {
			if (err) return callback(err);
			var match, regex = /(?:([ MADRCU?])([ MADRCU?]) )?(.*?)\0/g;
			var files = [];
			while ((match = regex.exec(stdout)) !== null) {
				files.push({
					staged: match[1] === undefined || match[1] !== ' ' && match[1] !== '?',
					unstaged: match[2] !== undefined && match[2] !== ' ',
					stagedStatusStr: me._getStagedFileStatusLabel(match[1], match[2]),
					unstagedStatusStr: me._getUnstagedFileStatusLabel(match[1], match[2]),
					name: match[3]
				});
			}
			callback(null, files);
		});
	},
	
	_getStagedFileStatusLabel: function(stagedStatus, unstagedStatus) {
		if (stagedStatus === undefined || stagedStatus === 'D') {
			return 'deleted';
		} else if (['A', '?', 'R', 'C'].indexOf(stagedStatus) > -1) {
			return 'new';
		} else if (['M', 'U'].indexOf(stagedStatus) > -1) {
			return 'modified';
		}
		return null;
	},
	
	_getUnstagedFileStatusLabel: function(stagedStatus, unstagedStatus) {
		if (unstagedStatus === 'D') {
			return 'deleted';
		} else if (unstagedStatus === 'A' || unstagedStatus === '?') {
			return 'new';
		} else if (['M', 'U'].indexOf(unstagedStatus) > -1) {
			return 'modified';
		}
		return null;
	}
};
