// git-watcher/GitWatcher.js
var fs = require('fs'),
	util = require('util'),
	async = require('async'),
	events = require('events'),
	ModuleWatcher = require('./ModuleWatcher');

var GitWatcher = function(baseGitRepoPath) {
	events.EventEmitter.call(this);
	this.path = baseGitRepoPath;
	this.modules = {};
	this.moduleNames = [];
};

util.inherits(GitWatcher, events.EventEmitter);

GitWatcher.prototype.init = function() {
	var me = this;
	me._initModule(me.path);
	me._findModules(me.path, function(err) {
		if (err) me.emit('error', err);
		else {
			me.moduleNames = Object.keys(me.modules);
			me.moduleNames.sort();
			me.emit('ready');
		}
	});
};

GitWatcher.prototype.getStatus = function(callback) {
	var me = this;
	var statuses = {};
	async.each(Object.keys(me.modules), function(module, callback) {
		me.modules[module].getStatus(function(err, status) {
			if (err) return callback(err);
			statuses[module] = status;
			callback();
		});
	}, function(err) {
		callback(err, statuses);
	});
};

GitWatcher.prototype.getModuleStatus = function(module, callback) {
	this.modules[module].getStatus(callback);
};

GitWatcher.prototype.getModules = function() {
	return this.moduleNames;
};

GitWatcher.prototype._initModule = function(fullPath) {
	var me = this;
	var module = new ModuleWatcher(fullPath);
	var moduleRelativeRepoName = fullPath.replace(require('path').dirname(me.path), '');
	module.init();
	module.on('change', function(status) {
		me.emit('change', {
			module: moduleRelativeRepoName,
			status: status
		});
	});
	module.on('error', function(err) {
		me.emit('error', err);
	});
	this.modules[moduleRelativeRepoName] = module;
};

GitWatcher.prototype._findModules = function(basePath, callback) {
	var me = this;
	fs.readdir(basePath, function(err, files) {
		if (err) return callback(err);
		async.each(files, function(fileName, callback) {
			if (fileName === '.git') return callback(); // do not watch .git stuff
			var fullPath = require('path').join(basePath, fileName);
			fs.stat(fullPath, function(err, stat) {
				if (err) return callback(err);
				if (stat.isDirectory()) {
					fs.exists(fullPath + '/.git', function(isSubmodule) {
						if (isSubmodule) {
							me._initModule(fullPath);
						}
						me._findModules(fullPath, callback);
					});
				} else {
					callback();
				}
			});
		}, callback);
	});
};

module.exports = GitWatcher;
